core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -436,20 +437,27 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
         }
     }
 
-    public IndexShard relocated(String reason) throws IndexShardNotStartedException {
-        try (Releasable block = suspendableRefContainer.blockAcquisition()) {
-            // no shard operation locks are being held here, move state from started to relocated
-            synchronized (mutex) {
-                if (state != IndexShardState.STARTED) {
-                    throw new IndexShardNotStartedException(shardId, state);
+    public void relocated(String reason) throws IllegalIndexShardStateException, InterruptedException {
+        try {
+            indexShardOperationsLock.blockOperations(30, TimeUnit.MINUTES, () -> {
```

Comment:
although I like it, it means the implementation on the indexShardOperationsLock can not use simple synchornized methods  because their duration is unclear . if this returned a releasble like it used to, I *think* all ops over there can be synchorinze and we won't need fancy non-blocking implementations with `while(true)` . If so, I think that's the right trade off to roll this back?